### PR TITLE
fix(helpers): add fill_input, wait_for_element, wait_for_network_idle

### LINKED ISSUE #257: Core helpers fail silently on modern web apps (React, Vue, Ember, SPAs)
### Problem

Two fundamental helpers don't work on the majority of modern sites:

**1. `type_text()` leaves submit buttons disabled**
It calls `Input.insertText` which bypasses framework event listeners. On React controlled inputs, Vue `v-model`, and Ember tracked properties, the field looks filled but the framework's internal state never updates — submit stays disabled.

**2. `wait_for_load()` returns before the page is usable on SPAs**
It polls `document.readyState == 'complete'`, which fires before any SPA framework renders content. Agents proceed to interact with elements that don't exist yet, causing silent failures or timeouts.

### Proposed additions to `helpers.py`

- `fill_input(selector, text)` — focuses element, types per-character via `press_key()`  (dispatches real `keyDown`/`char`/`keyUp`), then fires `input`+`change` events
- `wait_for_element(selector, timeout, visible)` — polls `querySelector` until found
- `wait_for_network_idle(timeout, idle_ms)` — uses existing `drain_events()` to detect  when all XHR/fetch activity has settled

No daemon changes needed.

### PR DESCRIPTION
**Pull request by wdeveloper16**
## Problem

Two core helpers silently fail on modern web apps:

**`type_text()` leaves submit buttons disabled on React / Vue / Ember** 
It calls `Input.insertText` which bypasses framework event listeners. The field looks filled but the framework's internal state never updates — submit stays disabled.

**`wait_for_load()` returns before SPAs are usable**
It polls `document.readyState == 'complete'`, which fires before any SPA framework renders content. 
Agents proceed to interact with elements that don't exist yet.

## Changes

Three new helpers in `helpers.py` — no daemon changes needed:

- **`fill_input(selector, text, clear_first=True)`** — focuses element, types  per-character via `press_key()` (dispatches real `keyDown`/`char`/`keyUp`),  then fires synthetic `input`+`change` events so React/Vue/Ember state updates
- **`wait_for_element(selector, timeout=10.0, visible=False)`** — polls  `querySelector` until element exists (and is visible if `visible=True`), returns `bool`
- **`wait_for_network_idle(timeout=10.0, idle_ms=500)`** — drains `drain_events()`  until no `Network.*` events arrive for `idle_ms` ms, useful after form submits  and SPA route transitions

## Tests

13 new unit tests covering all helpers and edge cases. 35 total, all passing.

## Closes: #257 

---
## Summary by cubic
Adds SPA-safe helpers to reliably fill inputs and wait for UI readiness. Fixes issues with React/Vue/Ember controlled inputs and premature interactions after navigation.

- **New Features**
  - `fill_input(selector, text, clear_first=True, timeout=0.0)` — focuses the input, optionally clears via Cmd/Ctrl+A, types per character with real key events, fires `input`/`change`, can wait for late-rendered elements, and raises if the element is missing.
  - `wait_for_element(selector, timeout=10.0, visible=False)` — polls until the element exists; if `visible=True`, uses `element.checkVisibility` with a computed-style fallback. Returns True/False.
  - `wait_for_network_idle(timeout=10.0, idle_ms=500)` — waits for no in-flight requests and no `Network.*` events for `idle_ms`. Useful after submits and SPA route changes.

- **Bug Fixes**
  - Dispatch select-all without a `char` event so Cmd/Ctrl+A reliably fires and clears inputs on all platforms.
  - Network idle detection tracks in-flight requests to prevent early returns.

<sup>Written for commit 4da0684eb2e4fa464338fb9e83b28159034e1b30. Summary will update on new commits.</sup>

### CODE REVIEW

File: src/browser_harness/helpers.py [hunk 1]
```diff
@@ -205,6 +205,26 @@ def click_at_xy(x, y, button="left", clicks=1):
...
+def fill_input(selector, text, clear_first=True):
+    """Fill a framework-managed input (React controlled, Vue v-model, Ember tracked).
+
+    type_text() uses Input.insertText which bypasses framework event listeners and leaves
+    submit buttons disabled. This helper focuses the element, clears it, types via real
+    key events, then fires synthetic input+change events so the framework sees the update.
+    """
+    js(f"document.querySelector({json.dumps(selector)})?.focus()")
+    if clear_first:
+        press_key("a", modifiers=2)  # Ctrl+A
```
[the remark this answers is not in the document]
Comment (wdeveloper16) [reply src/browser_harness/helpers.py]: Good catch. Fixed by detecting `sys.platform == "darwin"` and using `modifiers=4` (Cmd) on macOS, `modifiers=2` (Ctrl) elsewhere.

File: tests/unit/test_helpers.py [hunk 2]
```diff
@@ -50,3 +51,185 @@ def fake_cdp(method, **kwargs):
...
+
+    with patch("browser_harness.helpers._send", side_effect=fake_send), \
+         patch("browser_harness.helpers.time") as mock_time:
+        start = 1000.0
+        mock_time.time.side_effect = [start, start, start + 0.1, start + 0.1, start + 0.7, start + 0.7]
+        mock_time.sleep = lambda _: None
+        result = helpers.wait_for_network_idle(timeout=5.0, idle_ms=500)
+
+    assert result is True
+
```
[the remark this answers is not in the document]
Comment (wdeveloper16) [reply tests/unit/test_helpers.py]: Replaced with a 4-iteration test where `>idle_ms` elapses between `requestWillBeSent` and `loadingFinished`. 
The `assert idx == 4` ensures the function didn't short-circuit at iter2 — a silence-only implementation would fail this.

**Comment by Qodo-Free-For-OSS** [comment]
Hi, fill_input() uses optional-chaining focus and does not verify the selector matched; it will still send key events to whatever element is currently focused, causing input to land in the wrong place when the selector is missing or late-rendered.

**Severity:** action required | **Category:** correctness

**How to fix:** Validate element exists before typing

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> `fill_input()` can type into the wrong place when the selector doesn't match (or doesn't exist yet) because it never checks whether focusing succeeded.
> 
> ### Issue Context
> The function focuses via optional chaining and then immediately sends key events.
> 
> ### Fix Focus Areas
> - src/browser_harness/helpers.py[208-226]
>   - Make the focus step return a boolean (via `js(...)`) and raise/return `False` if not found.
>   - Optionally call `wait_for_element(selector, timeout=...)` before typing to support SPA rendering.
>   - Consider returning `True/False` to align with other wait helpers.

We noticed a couple of other issues in this PR as well - happy to share if helpful.

---
*[Qodo](https://github.com/marketplace/qodo-merge-pro-for-open-source) code review - free for open-source.*

**Comment by wdeveloper16** [comment]
Hi, @Alezander9 
Sorry to ping you.
Are you available to review the PR or are there any progress for my contribution? 🙂

**Comment by Alezander9** [comment]
Seems helpful, but why is your account so new? I am assuming this is a bot account?

**Comment by wdeveloper16** [comment]
Hi, @Alezander9  thanks for your response.
I’m not a bot 🙂
I’ve actually been working at one company for a long time,
Now I have more free time and would love to contribute to the browser-use as a contributor, I’m a pretty passionate coder.
Anyway, I'd appreciate your feedback with my contributions.
Thanks

**Comment by sauravpanda** [comment]
Pushed a small tweak to `wait_for_element(visible=True)`: it now prefers `el.checkVisibility({checkOpacity:true, checkVisibilityCSS:true})` and falls back to the existing per-element `getComputedStyle` check on older Chrome. Reason: a computed-style check on the element alone returns the descendant's own non-`none` value even when an ancestor has `display:none`, so the helper would report nested-but-hidden elements as visible.

The earlier cubic review on this PR predates the current state and lists three findings (Cmd+A, CSS visibility check, network in-flight tracking) that are already implemented at HEAD — confirmed by reading the code on the latest commit. @cubic-dev-ai please re-run on the latest commit so the review reflects current state.

**Comment by sauravpanda** [comment]
Cubic's P1 here is correct — I was wrong to dismiss the earlier round on the same area. Even with the platform-correct modifier (`Cmd` on darwin, `Ctrl` elsewhere), `press_key` always emits a follow-up `Input.dispatchKeyEvent` of type `char` for any single-character key. With the modifier held, that `char` event makes Chrome treat the input as a printable letter `"a"` rather than firing the select-all shortcut, so the field never actually gets cleared (Backspace then deletes the stray `"a"` and leaves the original value untouched).

Pushed `4da0684` to address it:

- `fill_input` now dispatches the select-all `rawKeyDown`/`keyUp` pair directly, with `modifiers` set and no `text`/`char` event.
- Tightened `test_fill_input_clear_first_*` to assert (a) the `"a"` event carries the platform-correct modifier (`4` on darwin, `2` elsewhere) and (b) no `char` event with `text="a"` is emitted.

@cubic-dev-ai please re-run on the latest commit.

**Comment by wdeveloper16** [comment]
Thanks for your review, @sauravpanda
